release/train_tests/benchmark/dataloader_factory.py

```diff
@@ -36,8 +521,268 @@ def get_ray_datasets(self) -> Dict[str, Dataset]:
+                    current_time = time.time()
+                    time_since_last_batch = current_time - last_batch_time
+                    if time_since_last_batch > 10:
+                        logger.warning(
+                            f"[DataLoader] Worker {worker_rank}: "
+                            f"Long delay ({time_since_last_batch:.2f}s) between batches {batch_idx-1} and {batch_idx}"
+                        )
+
+                    # Move batch to device
+                    images, labels = batch
+                    logger.info(
+                        f"[DataLoader] Worker {worker_rank} processing batch {batch_idx} "
+                        f"(shape: {images.shape}, time since last: {time_since_last_batch:.2f}s)"
+                    )
+
+                    transfer_start = time.time()
+                    images = images.to(
+                        device, non_blocking=self.benchmark_config.torch_non_blocking
+                    )
+                    labels = labels.to(
+                        device, non_blocking=self.benchmark_config.torch_non_blocking
+                    )
+                    transfer_time = time.time() - transfer_start
+
+                    if transfer_time > 5:
+                        logger.warning(
+                            f"[DataLoader] Worker {worker_rank}: "
+                            f"Slow device transfer ({transfer_time:.2f}s) for batch {batch_idx}"
+                        )
+
+                    logger.info(
+                        f"[DataLoader] Worker {worker_rank} completed device transfer "
+                        f"for batch {batch_idx} in {transfer_time:.2f}s"
+                    )
+
+                    last_batch_time = time.time()
+                    yield images, labels
+
+                except Exception as e:
+                    logger.error(
+                        f"[DataLoader] Worker {worker_rank} error processing batch {batch_idx}: {str(e)}",
+                        exc_info=True,
+                    )
+                    raise
+
+        except Exception as e:
+            logger.error(
+                f"[DataLoader] Worker {worker_rank} error in batch iterator: {str(e)}",
+                exc_info=True,
+            )
+            raise
+
+    def get_train_dataloader(self) -> Iterator[Tuple[torch.Tensor, torch.Tensor]]:
+        """Create a DataLoader for training data.
+
+        Returns:
+            An iterator that yields (image, label) tensors for training
+        """
+        worker_rank = ray.train.get_context().get_world_rank()
+        logger.info(f"[DataLoader] Worker {worker_rank} creating train dataloader")
+
+        dataloader_config = self.get_dataloader_config()
+        device = self._get_device()
+
+        # Create dataset and dataloader
+        train_ds = S3ParquetImageIterableDataset(
+            file_urls=self._get_file_urls(self.train_url),
+            random_transforms=True,
+        )
```

Comment:
We can make this `TorchDataloaderFactory` class agnostic to the actual dataset by making this `train_ds` initialization happen in an abstract method `get_iterable_datasets`, similar to this:

https://github.com/ray-project/ray/blob/a3d571920ed181d25c281fbaf5e83615f9383347/release/train_tests/benchmark/dataloader_factory.py#L52-L59

Then the dataloader factory just wraps the underlying dataset object with the right settings.